src/include/processor/operator/persistent/copy_node.h

```diff
@@ -35,22 +35,27 @@ class CopyNodeSharedState {
 private:
     inline common::offset_t getNextNodeGroupIdxWithoutLock() { return currentNodeGroupIdx++; }
 
+    void calculateNumTuples();
```

Comment:
Ideally, you don't need to calculate num tuples. Statistics can provide the information for you after `finalize`.

Reply:
hmmm you do need to calculate in order to update Statistics. This is a refactor of existing code.